File: leetcode/medium/114-flatten-bt-linked/iterative.py
"""
  @author Vortexx2

  Runtime - 36 ms O(n)
  Memory Usage - 15.2 MB O(1)
"""
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def flatten(self, root: Optional[TreeNode]) -> None:
    """
    Do not return anything, modify root in-place instead.
    """

    while root:
      if root.left:
        pre = root.left
        while pre.right:
          pre = pre.right

        pre.right = root.right
        root.right = root.left
        root.left = None

      root = root.right

File: leetcode/medium/114-flatten-bt-linked/iterative.cpp
/**
 * @file iterative.cpp
 * @author Vortexx2
 * @brief Problem 114 - Flatten BT to Linked List
 * @date 22-08-2021
 *
 * Runtime - 4 ms O(n)
 * Memory Usage - 12.7 MB O(1)
 */

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  void flatten(TreeNode *root) {
    TreeNode *node = root;
    while (node) {
      if (node->left) {
        TreeNode *pre = node->left;
        while (pre->right) pre = pre->right;

        pre->right = node->right;
        node->right = node->left;
        node->left = nullptr;
      }
      node = node->right;
    }
  }
};